src/device/device-management.ts

```diff
@@ -33,29 +33,87 @@ export type AudioDeviceConstraints = {
   deviceId?: string;
 };
 
-export type VideoDeviceConstraints = {
-  deviceId?: ConstrainDOMString;
-  width?: ConstrainULong;
-  height?: ConstrainULong;
-  aspectRatio?: ConstrainDouble;
-  frameRate?: ConstrainDouble;
-  facingMode?: ConstrainDOMString;
+// Predefined encoder config mentioned similar to https://api-ref.agora.io/en/voice-sdk/web/4.x/globals.html#videoencoderconfigurationpreset
+export enum StaticVideoEncoderConfig {
+  Profile1080P = '1080P',
+}
+
+export type ConstrainLong = {
+  min?: number;
+  max?: number;
+  ideal?: number;
+  exact?: number;
+};
+
+// Predefined encoder config mentioned similar to https://api-ref.agora.io/en/voice-sdk/web/4.x/globals.html#videoencoderconfigurationpreset
+export type videoEncoderConfig = {
+  bitRateMin?: number;
+  bitrateMax?: number;
+  frameRate?: number | ConstrainLong;
+  height?: number | ConstrainLong;
+  width?: number | ConstrainLong;
+  aspectRatio?: number;
+};
+
+export type VideoConstraints = {
+  cameraDeviceId?: string;
+  faceingMode?: ConstrainULong;
+  optimizationMode?: ConstrainULong;
+  encoderConfig?: StaticVideoEncoderConfig | videoEncoderConfig;
+};
+
+// something like this https://api-ref.agora.io/en/video-sdk/web/4.x/globals.html#audioencoderconfigurationpreset
+export enum StaticAudioEncoderConfig {
+  sampleRateLow = '16kHz',
+}
+
+export type audioEncoderConfig = {
+  bitrate?: number;
+  sampleRate?: number;
+  sampleSize?: number;
+  stereo?: boolean;
 };
 
+export type MicrophoneConstraints = {
+  echoCancelation?: boolean;
+  autoGainControl?: boolean;
+  noiseSupression?: boolean;
+  microphoneDeviceid?: string;
+  encoderConfig?: StaticAudioEncoderConfig | audioEncoderConfig;
+};
+
+export type CustomVideoTrackInitConfig = {
+  mediaStreamTrack: MediaStreamTrack;
+  encoderConfig?: StaticVideoEncoderConfig | videoEncoderConfig;
+};
+
+export type CustomAudioTrackInitConfig = {
+  mediaStreamTrack: MediaStreamTrack;
+  encoderConfig?: StaticAudioEncoderConfig | audioEncoderConfig;
+};
+
+// /**
+//  * @param audioConfig
+//  * @param videoConfig
+//  */
+// export async function createMicrophoneAndCameraTracks(
+//   audioConfig?: MicrophoneConstraints,
+//   videoConfig?: VideoConstraints
+// ): Promise<[IMicrophoneAudioTrack, ICameraVideoTrack]> {}
+
 /**
  * Creates a camera video track.
  *
  * @param constraints - Video device constraints.
  * @returns A LocalTrack object or an error.
  */
-export async function createCameraTrack(
-  constraints?: VideoDeviceConstraints
-): Promise<LocalCameraTrack> {
+export async function createCameraTrack(constraints?: VideoConstraints): Promise<LocalCameraTrack> {
   let stream: MediaStream;
   try {
-    stream = await media.getUserMedia({ video: { ...constraints } });
+    // TODO: change it
+    stream = await media.getUserMedia({ video: {} });
```

Comment:
This should be using `constraints`?

Reply:
yes we need to generate the constraints , @haohancom  is going to add it